src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+    {
+        ['id' => $id, 'version' => $version, 'content' => $content, 'hash' => $hash] = $row;
+
+        // Some drivers, eg pdo_sqlite, unfortunately return strings rather than resources.
+        // Doctrine doesn't provide an equivalent of `PDOStatement::bindColumn()` either.
+        // This might cause memory problems if the content is too large.
+        $content = Type::getType('blob')->convertToPHPValue($content, $this->connection->getDatabasePlatform());
+
+        return new ItemVersion(ItemId::fromString($id), ItemVersionNumber::create($version), $content, $hash);
+    }
+
+    private function getCurrentVersion(ItemId $id, int $lockMode = LockMode::NONE) : ItemVersionNumber
+    {
+        $platform = $this->connection->getDatabasePlatform();
+
+        $sql = 'SELECT item.version FROM %s WHERE item.id = :id %s';
+        if (LockMode::NONE === $lockMode) {
+            $sql = sprintf($sql, "{$this->tableItems} item", '');
+        } else {
+            $sql = sprintf(
+                $sql,
+                $platform->appendLockHint("{$this->tableItems} item", LockMode::PESSIMISTIC_READ),
+                $platform->getWriteLockSQL()
+            );
+        }
+
+        $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id]);
+        $current = $stmt->fetch(FetchMode::COLUMN);
+
+        if (false === $current) {
+            throw new ItemNotFound($id);
+        }
+
+        return ItemVersionNumber::create($current);
+    }
+
+    private function deleteItem(ItemId $id) : void
+    {
+        $this->connection->delete($this->tableItems, ['id' => (string) $id]);
+        $this->connection->delete($this->tableVersions, ['id' => (string) $id]);
+    }
+
+    private function deleteVersion(ItemId $id, ItemVersionNumber $current) : void
+    {
+        $this->connection->update(
+            $this->tableItems,
+            ['version' => $current->previous()],
+            ['id' => (string) $id]
+        );
+
+        $this->connection->delete(
+            $this->tableVersions,
+            ['id' => (string) $id, 'version' => $current->toInt()]
+        );
+    }
+
+    private function insertNewItem(ItemVersion $item) : void
+    {
+        $this->connection->insert(
+            $this->tableItems,
+            [
+                'id' => (string) $item->getId(),
+                'version' => $item->getVersion()->toInt(),
+            ]
+        );
+
+        rewind($item->getContent());
+
+        $this->connection->insert(
+            $this->tableVersions,
+            [
+                'id' => (string) $item->getId(),
+                'version' => $item->getVersion()->toInt(),
+                'content' => $item->getContent(),
+                'hash' => $item->getHash(),
+            ],
+            ['content' => ParameterType::LARGE_OBJECT]
+        );
+    }
+
+    private function insertVersion(ItemVersion $item) : void
+    {
+        $this->connection->update(
```

Comment:
it's all transactional but isn't the natural order to add an invisible version, than modify the current version field to point to it?